From: Folia Compatibility Patch
Subject: [PATCH 0004] ExecutionTask: Remove SchedulerUtils and BukkitRunnable, migrate to Folia

AE PATCH REASON: SchedulerUtils and BukkitRunnable are incompatible with Folia server architecture.
AE PATCH FIX: Remove SchedulerUtils and BukkitRunnable imports. Replace BukkitRunnable with Folia RegionScheduler.runDelayed() and EntityScheduler.runDelayed(). Remove @Generated lombok annotations. Rename obfuscated variables to meaningful names throughout.
AE PATCH RISK: LOW

---
diff --git a/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java b/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java
--- a/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java	2026-07-30 10:27:31.670350141 +0000
+++ b/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java	2026-07-30 10:28:05.843658221 +0000
@@ -4,7 +4,6 @@
 import java.util.Arrays;
 import java.util.LinkedList;
 import java.util.List;
-import lombok.Generated;
 import net.advancedplugins.ae.impl.effects.api.EffectsActivateEvent;
 import net.advancedplugins.ae.impl.effects.api.EffectsActivatedEvent;
 import net.advancedplugins.ae.impl.effects.armorutils.ArmorType;
@@ -20,7 +19,6 @@
 import net.advancedplugins.ae.impl.effects.effects.variables.Variables;
 import net.advancedplugins.ae.impl.utils.ASManager;
 import net.advancedplugins.ae.impl.utils.LocalLocation;
-import net.advancedplugins.ae.impl.utils.SchedulerUtils;
 import net.advancedplugins.ae.impl.utils.hooks.HookPlugin;
 import net.advancedplugins.ae.impl.utils.hooks.HooksHandler;
 import net.advancedplugins.ae.impl.utils.hooks.plugins.MythicMobsHook;
@@ -31,7 +29,6 @@
 import org.bukkit.entity.Player;
 import org.bukkit.event.block.BlockBreakEvent;
 import org.bukkit.event.entity.EntityDamageByEntityEvent;
-import org.bukkit.scheduler.BukkitRunnable;
 
 public class ExecutionTask {
    private ActionExecution actionExecution;
@@ -46,229 +43,232 @@
    private TargetResults targetResults;
    private final DamageHandler damageHandler = new DamageHandler();
 
-   public ExecutionTask(AdvancedAbility var1, ActionExecution var2, ActionExecutionBuilder var3) {
-      this.ability = var1;
-      this.actionExecution = var2;
-      this.builder = var3;
-      this.location = var3.getMain().getLocation();
+   public ExecutionTask(AdvancedAbility ability, ActionExecution actionExecution, ActionExecutionBuilder builder) {
+      this.ability = ability;
+      this.actionExecution = actionExecution;
+      this.builder = builder;
+      this.location = builder.getMain().getLocation();
    }
 
    public void init() {
-      if (!this.soulboundOnly || !this.actionExecution.getAllEffects().stream().noneMatch(var0 -> var0.equalsIgnoreCase("KEEP_ON_DEATH"))) {
-         ArrayList var1 = new ArrayList();
+      if (!this.soulboundOnly || !this.actionExecution.getAllEffects().stream().noneMatch(effect -> effect.equalsIgnoreCase("KEEP_ON_DEATH"))) {
+         ArrayList otherTargets = new ArrayList();
          if (!this.sets && Bukkit.isPrimaryThread()) {
-            EffectsActivateEvent var2 = new EffectsActivateEvent(this.ability, this.builder.getMain(), this.builder.getOther(), this);
-            Bukkit.getPluginManager().callEvent(var2);
-            if (this.ability.getEffects().isEmpty() || var2.isCancelled()) {
+            EffectsActivateEvent activateEvent = new EffectsActivateEvent(this.ability, this.builder.getMain(), this.builder.getOther(), this);
+            Bukkit.getPluginManager().callEvent(activateEvent);
+            if (this.ability.getEffects().isEmpty() || activateEvent.isCancelled()) {
                return;
             }
 
-            if (var2.getOtherTargets() != null) {
-               var1.addAll(var2.getOtherTargets());
+            if (activateEvent.getOtherTargets() != null) {
+               otherTargets.addAll(activateEvent.getOtherTargets());
             }
          }
 
-         for (final String var3 : this.ability.getEffects()) {
-            final ArrayList var4 = new ArrayList();
-            final LinkedList var5 = new LinkedList();
-            var3 = this.actionExecution.parseVariables(var3);
-            this.updateMain(var5, var4, var3);
-            var5.addAll(var1);
-            if (var3.startsWith("WAIT")) {
+         for (String effect : this.ability.getEffects()) {
+            final List<Location> locations = new ArrayList<>();
+            final LinkedList<LivingEntity> entities = new LinkedList<>();
+            String parsedEffect = this.actionExecution.parseVariables(effect);
+            this.updateMain(entities, locations, parsedEffect);
+            entities.addAll(otherTargets);
+            if (parsedEffect.startsWith("WAIT")) {
                this.wait = true;
             }
 
             if (this.targetResults.getEffect() != null) {
-               var3 = this.actionExecution.parseVariables(this.targetResults.getEffect());
+               parsedEffect = this.actionExecution.parseVariables(this.targetResults.getEffect());
             }
 
-            String[] var6 = var3.split(":");
-            if (var3.startsWith("WAIT")) {
-               this.currentWait = this.currentWait + ASManager.parseInt(var6[1], 0);
+            String[] parts = parsedEffect.split(":");
+            if (parsedEffect.startsWith("WAIT")) {
+               this.currentWait = this.currentWait + ASManager.parseInt(parts[1], 0);
             } else if (!this.wait) {
-               if (!var4.isEmpty()) {
-                  for (Location var12 : var4) {
-                     this.activate(var3, this.builder.getMain(), var12);
+               if (!locations.isEmpty()) {
+                  for (Location location : locations) {
+                     this.activate(parsedEffect, this.builder.getMain(), location);
                   }
                } else {
-                  for (LivingEntity var8 : var5) {
-                     this.activate(var3, var8, null);
+                  for (LivingEntity entity : entities) {
+                     this.activate(parsedEffect, entity, null);
                   }
                }
             } else {
-               (new BukkitRunnable() {
-                  public void run() {
-                     ExecutionTask.this.builder.setDamageEventNotGoingToRun(true);
-                     if (!var4.isEmpty()) {
-                        for (Location var2 : var4) {
-                           ExecutionTask.this.activate(var3, ExecutionTask.this.builder.getMain(), var2);
-                        }
-                     } else {
-                        for (LivingEntity var4x : var5) {
-                           ExecutionTask.this.activate(var3, var4x, null);
-                        }
-                     }
+               final String finalEffect = parsedEffect;
+               final List<Location> finalLocs = new ArrayList<>(locations);
+               final List<LivingEntity> finalEnts = new ArrayList<>(entities);
+               long delay = (long)this.currentWait < 1L ? 1L : (long)this.currentWait;
+               ExecutionTask self = this;
+               if (!finalLocs.isEmpty()) {
+                  for (Location loc : finalLocs) {
+                     Bukkit.getRegionScheduler().runDelayed(ASManager.getInstance(), loc, task2 -> {
+                        self.builder.setDamageEventNotGoingToRun(true);
+                        self.activate(finalEffect, self.builder.getMain(), loc);
+                     }, delay);
+                  }
+               } else {
+                  for (LivingEntity ent : finalEnts) {
+                     ent.getScheduler().runDelayed(ASManager.getInstance(), task2 -> {
+                        self.builder.setDamageEventNotGoingToRun(true);
+                        self.activate(finalEffect, ent, null);
+                     }, () -> {}, delay);
                   }
-               }).runTaskLater(EffectsHandler.getInstance(), (long)this.currentWait);
+               }
             }
          }
       }
    }
 
-   private void activate(String var1, LivingEntity var2, Location var3) {
-      ASManager.debug("§d[ExecutionTask] start of " + var1 + " parsing+");
-      String var4 = DynamicVariable.parseThroughCustomVariables(
-         Variables.replaceVariables(var1, this.builder.getAttacker(), this.builder.getVictim(), this.actionExecution)
+   private void activate(String effect, LivingEntity target, Location location) {
+      ASManager.debug("§d[ExecutionTask] start of " + effect + " parsing+");
+      String parsedEffect = DynamicVariable.parseThroughCustomVariables(
+         Variables.replaceVariables(effect, this.builder.getAttacker(), this.builder.getVictim(), this.actionExecution)
       );
-      var4 = EffectsHandler.getVariablesHandler().parseEffectLine(var4, this.builder.getMain(), this);
-      var4 = EffectsHandler.getFunctionsHandler().parseEffectLine(var4, this.builder.getMain(), this);
-      var4 = EffectsHandler.getPointersHandler().parseEffectLine(var4, this.builder.getType(), this);
-      if (!var4.isEmpty() && !var4.contains("$skip")) {
-         String[] var5 = var4.replaceAll(" ", "").split(":");
-         String[] var6 = Arrays.copyOfRange(var5, 1, var5.length);
-         AdvancedEffect var7 = EffectsHandler.getEffectStorage().getEffect(var5[0].replaceAll(" ", ""));
-         if (var7 == null) {
+      parsedEffect = EffectsHandler.getVariablesHandler().parseEffectLine(parsedEffect, this.builder.getMain(), this);
+      parsedEffect = EffectsHandler.getFunctionsHandler().parseEffectLine(parsedEffect, this.builder.getMain(), this);
+      parsedEffect = EffectsHandler.getPointersHandler().parseEffectLine(parsedEffect, this.builder.getType(), this);
+      if (!parsedEffect.isEmpty() && !parsedEffect.contains("$skip")) {
+         String[] parts = parsedEffect.replaceAll(" ", "").split(":");
+         String[] effectArgs = Arrays.copyOfRange(parts, 1, parts.length);
+         AdvancedEffect advancedEffect = EffectsHandler.getEffectStorage().getEffect(parts[0].replaceAll(" ", ""));
+         if (advancedEffect == null) {
             this.reportIssue(
-               var4,
-               "Failed to activate effects as advancedEffect is null or invalid: '" + var5[0].replaceAll(" ", "") + "'",
-               var1,
+               parsedEffect,
+               "Failed to activate effects as advancedEffect is null or invalid: '" + parts[0].replaceAll(" ", "") + "'",
+               effect,
                "trigger:" + this.builder.getType(),
                "entity:" + this.builder.getMain().getName()
             );
          } else {
-            if (var7.hasStringArgument()) {
-               var5 = var4.split(":");
-               var6 = Arrays.copyOfRange(var5, 1, var5.length);
+            if (advancedEffect.hasStringArgument()) {
+               parts = parsedEffect.split(":");
+               effectArgs = Arrays.copyOfRange(parts, 1, parts.length);
             }
 
-            if (var7.isBlockEffect() && var3 == null && this.builder.getEvent() instanceof BlockBreakEvent) {
-               var3 = this.builder.getBlock().getLocation();
+            if (advancedEffect.isBlockEffect() && location == null && this.builder.getEvent() instanceof BlockBreakEvent) {
+               location = this.builder.getBlock().getLocation();
             }
 
-            if (this.builder.getMain() instanceof Player && !this.builder.getMain().isOp() && var7.isExemptFromAC()) {
-               EffectsHandler.getAntiCheatHooks().forEach((var1x, var2x) -> var2x.exempt((Player)this.builder.getMain()));
+            if (this.builder.getMain() instanceof Player && !this.builder.getMain().isOp() && advancedEffect.isExemptFromAC()) {
+               EffectsHandler.getAntiCheatHooks().forEach((hookName, hook) -> hook.exempt((Player)this.builder.getMain()));
             }
 
             if (this.builder.getEvent() instanceof EntityDamageByEntityEvent) {
-               EntityDamageByEntityEvent var8 = (EntityDamageByEntityEvent)this.builder.getEvent();
-               if (HooksHandler.isEnabled(HookPlugin.MYTHICMOBS) && MythicMobsHook.getIgnoreEnchantsMobs().contains(var8.getDamager())) {
+               EntityDamageByEntityEvent damageEvent = (EntityDamageByEntityEvent)this.builder.getEvent();
+               if (HooksHandler.isEnabled(HookPlugin.MYTHICMOBS) && MythicMobsHook.getIgnoreEnchantsMobs().contains(damageEvent.getDamager())) {
                   return;
                }
             }
 
             ASManager.debug(
                "§d§l(!) "
-                  + var4
+                  + parsedEffect
                   + " activating "
-                  + (var3 != null ? "location based " + new LocalLocation(var3).getEncode() : "entity based " + (var2 != null ? var2.getType().name() : "null"))
+                  + (location != null ? "location based " + new LocalLocation(location).getEncode() : "entity based " + (target != null ? target.getType().name() : "null"))
             );
 
             try {
-               if (var3 != null) {
-                  if (var7.executeEffect(this, var3, var6)) {
+               if (location != null) {
+                  if (advancedEffect.executeEffect(this, location, effectArgs)) {
                      if (!Bukkit.isPrimaryThread()) {
-                        SchedulerUtils.runTask(() -> {
-                           EffectsActivatedEvent var2x = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                           Bukkit.getPluginManager().callEvent(var2x);
+                        final Location locFinal = location;
+                        Bukkit.getRegionScheduler().execute(ASManager.getInstance(), locFinal, () -> {
+                           EffectsActivatedEvent evt = new EffectsActivatedEvent(this.ability, this.builder.getMain(), target, this);
+                           Bukkit.getPluginManager().callEvent(evt);
                         });
                      } else {
-                        EffectsActivatedEvent var14 = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                        Bukkit.getPluginManager().callEvent(var14);
+                        EffectsActivatedEvent activatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), target, this);
+                        Bukkit.getPluginManager().callEvent(activatedEvent);
                      }
                   } else {
-                     this.reportIssue(var4, "activating entity-based effect without a valid target (e.g. @Attacker)", this.ability.getName());
+                     this.reportIssue(parsedEffect, "activating entity-based effect without a valid target (e.g. @Attacker)", this.ability.getName());
                   }
-               } else if (!var7.executeEffect(this, var2, var6)) {
-                  this.reportIssue(var4, "activating location-based effect without a valid target (e.g. @Block)", this.ability.getName());
+               } else if (!advancedEffect.executeEffect(this, target, effectArgs)) {
+                  this.reportIssue(parsedEffect, "activating location-based effect without a valid target (e.g. @Block)", this.ability.getName());
                } else {
                   if (!Bukkit.isPrimaryThread()) {
-                     SchedulerUtils.runTask(() -> {
-                        EffectsActivatedEvent var2x = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                        Bukkit.getPluginManager().callEvent(var2x);
-                     });
+                     LivingEntity targetEnt = target != null ? target : this.builder.getMain();
+                     targetEnt.getScheduler().run(ASManager.getInstance(), task2 -> {
+                        EffectsActivatedEvent evt = new EffectsActivatedEvent(this.ability, this.builder.getMain(), target, this);
+                        Bukkit.getPluginManager().callEvent(evt);
+                     }, null);
                   } else {
-                     EffectsActivatedEvent var15 = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                     Bukkit.getPluginManager().callEvent(var15);
+                     EffectsActivatedEvent activatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), target, this);
+                     Bukkit.getPluginManager().callEvent(activatedEvent);
                   }
 
                   if (this.getBuilder().isPermanent() && this.getBuilder().isRemoved()) {
-                     SchedulerUtils.runTaskLater(
-                        () -> {
-                           if (var2 != null && !var2.isDead() && ASManager.isOnline(var2)) {
-                              ArmorType.getArmorContents(var2)
-                                 .forEach((var1x, var2x) -> ArmorWearTrigger.getArmorWearTrigger().runCheck(var2, var2x, var1x, false, false));
+                     if (target != null) {
+                        target.getScheduler().runDelayed(ASManager.getInstance(), task2 -> {
+                           if (!target.isDead() && ASManager.isOnline(target)) {
+                              ArmorType.getArmorContents(target)
+                                 .forEach((armorType, itemStack) -> ArmorWearTrigger.getArmorWearTrigger().runCheck(target, itemStack, armorType, false, false));
                            }
-                        },
-                        2L
-                     );
+                        }, () -> {}, 2L);
+                     }
                   }
                }
-            } catch (Exception var9) {
-               var9.printStackTrace();
-               this.reportIssue(var4, "Error while activating effect");
+            } catch (Exception e) {
+               e.printStackTrace();
+               this.reportIssue(parsedEffect, "Error while activating effect");
             }
          }
       }
    }
 
-   public void reportIssue(String var1, String... var2) {
-      for (Player var4 : Bukkit.getOnlinePlayers()) {
-         if (var4.hasPermission("ae.admin")) {
-            var4.sendMessage(Text.modify("&4Failed to activate effect '&f" + var1 + "'"));
-            var4.sendMessage(Text.modify("&cAdditional information: &7" + String.join(", ", var2)));
+   public void reportIssue(String effect, String... info) {
+      for (Player player : Bukkit.getOnlinePlayers()) {
+         if (player.hasPermission("ae.admin")) {
+            player.sendMessage(Text.modify("&4Failed to activate effect '&f" + effect + "'"));
+            player.sendMessage(Text.modify("&cAdditional information: &7" + String.join(", ", info)));
          }
       }
 
-      ASManager.getInstance().getLogger().warning(Text.modify("&4Failed to activate effect '&f" + var1 + "'"));
-      ASManager.getInstance().getLogger().warning(Text.modify("&cAdditional information: &7" + String.join(", ", var2)));
+      ASManager.getInstance().getLogger().warning(Text.modify("&4Failed to activate effect '&f" + effect + "'"));
+      ASManager.getInstance().getLogger().warning(Text.modify("&cAdditional information: &7" + String.join(", ", info)));
    }
 
-   private void updateMain(LinkedList<LivingEntity> var1, List<Location> var2, String var3) {
-      var1.clear();
-      var2.clear();
-      this.targetResults = EffectsHandler.getTargetHandler().handleTargets(var3, this);
+   private void updateMain(LinkedList<LivingEntity> entities, List<Location> locations, String effect) {
+      entities.clear();
+      locations.clear();
+      this.targetResults = EffectsHandler.getTargetHandler().handleTargets(effect, this);
       if (this.targetResults.getTargetList() != null) {
-         var1.addAll(this.targetResults.getTargetList());
+         entities.addAll(this.targetResults.getTargetList());
       }
 
       if (this.targetResults.getTargetLocations() != null) {
-         var2.addAll(this.targetResults.getTargetLocations());
+         locations.addAll(this.targetResults.getTargetLocations());
       }
 
-      var2 = ASManager.removeDuplicateLocations(var2);
+      locations = ASManager.removeDuplicateLocations(locations);
    }
 
-   public ExecutionTask asSets(boolean var1) {
-      this.sets = var1;
+   public ExecutionTask asSets(boolean sets) {
+      this.sets = sets;
       return this;
    }
 
-   public ExecutionTask setLocation(Location var1) {
-      this.location = var1;
+   public ExecutionTask setLocation(Location location) {
+      this.location = location;
       return this;
    }
 
-   public ExecutionTask soulboundOnly(boolean var1) {
-      this.soulboundOnly = var1;
+   public ExecutionTask soulboundOnly(boolean soulboundOnly) {
+      this.soulboundOnly = soulboundOnly;
       return this;
    }
 
-   @Generated
    public ActionExecution getActionExecution() {
       return this.actionExecution;
    }
 
-   @Generated
    public ActionExecutionBuilder getBuilder() {
       return this.builder;
    }
 
-   @Generated
    public AdvancedAbility getAbility() {
       return this.ability;
    }
 
-   @Generated
    public DamageHandler getDamageHandler() {
       return this.damageHandler;
    }
